internal/pkg/cli/deploy/workload.go

```diff
@@ -430,17 +468,20 @@ func (d *workloadDeployer) uploadContainerImages(out *UploadArtifactsOutput) err
 		pr, pw := io.Pipe()
 		g.Go(func() error {
 			defer pw.Close()
-			digest, err := d.repository.BuildAndPush(ctx, buildArgs, pw)
+			digest, err := in.BuildFunc(ctx, buildArgs, pw)
 			if err != nil {
 				return fmt.Errorf("build and push the image %q: %w", name, err)
 			}
 			digestsMu.Lock()
 			defer digestsMu.Unlock()
-			out.ImageDigests[name] = ContainerImageIdentifier{
+			in.Out.ImageDigests[name] = ContainerImageIdentifier{
 				Digest:            digest,
-				CustomTag:         d.image.CustomTag,
-				GitShortCommitTag: d.image.GitShortCommitTag,
+				CustomTag:         in.CustomTag,
+				GitShortCommitTag: in.GitShortCommitTag,
 			}
+			imageName := fmt.Sprintf("%s:%s", uri, buildArgs.Tags[0])
```

Comment:
do we validate anywhere that there is at least one tag?

Reply:
When we appending latest as tag to every container. So we have at least one tag. https://github.com/aws/copilot-cli/blob/123b8afbae354008a87b786723da002ff84303e1/internal/pkg/cli/deploy/workload.go#L492